content/kb/troubleshooting-failure-to-delete-orphaned-pod-volume-directory.md

```diff
@@ -0,0 +1,40 @@
+---
+title: "Troubleshooting: Failure to delete orphaned Pod volume directory"
+author: Ray Chang
+draft: false
+date: 2023-03-23
+categories:
+  - "csi"
+---
+
+## Applicable versions
+
+All Longhorn versions.
+Kubernetes versions before v1.27.
+
+## Symptoms
+
+In the event of a worker node failure, while hosting active Pods, the Pods are gracefully evicted as the node undergoes downtime and awaits restoration. During this period, the kubelet, which is responsible for managing the node, will generate the following error messages at regular intervals of two seconds.
+```
+orphaned pod <pod-uid> found, but error not a directory occurred when trying to remove the volumes dir
+```
+
+## Reason
+
+This situation occurs when a node goes through a downtime state and then takes some time before entering the recovery phase. During this process, the affected Pods are evicted and relocated to other nodes. However, due to the disruption, the connection between the kubelet and the longhorn-csi-plugin is severed. Therefore, the kubelet encountered difficulties when deleting the `vol_data.json` file. This process is used to perform self-housekeeping tasks to clean up orphan volume mount points associated with evicted Pods. The kubelet, while capable of removing directories, cannot delete individual files, resulting in incomplete cleanup in this specific situation. [(source code)](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_volumes.go#L155-L162).
+
+## Solution
+
+Once the node and kubelet have been restored, the longhorn-csi-plugin will automatically restart, allowing the Pod to remount the volume and resume its running state.
+
+However, in cases where the Pod and its associated volume are rescheduled to a different node, leaving behind a lingering `vol_data.json` file on the crashed node, manual intervention is required. You will need to manually delete the `vol_data.json` file located within the `/var/lib/kubelet/pods/<pod-uid>/volumes/kubernetes.io~csi/pvc_<pod-uid>/` directory.
+
+For Kubernetes versions 1.27.0 and above, this issue is addressed, ensuring that orphaned Pod volume mount points are properly cleaned up within the reconciliation loop.
```

Comment:
The backporting fix PR for v1.27 is not merged yet. Can we say this now?

Reply:
Yes, the [backport PR for v1.27](https://github.com/kubernetes/kubernetes/pull/117235) has not been merged yet. I would like to consult whether it would be more suitable if I modify it like this:

Within the present Kubernetes master branch, the issue is addressed, thereby ensuring that orphaned Pod volume mount points are properly cleaned up within the reconciliation loop. Moreover, a PR addressing the issue has been backported to version `1.27` and is presently awaiting the merging process.

Reply:
A good point. Maybe this makes my request to update the Applicable versions section premature.

Reply:
Thanks. I updated this paragraph as well as the "Applicable Versions" section.